### PR TITLE
Request access page

### PR DESCRIPTION
**Pull request by anissa-agahchen**
# Overview

## Links to jira tickets

BHBC-886 - IDIR Authenticated User Requests Access to BioHub {https://quartech.atlassian.net/browse/BHBC-886}

## This PR contains the following changes

- Access request page with submission to database
- number of auth related changes

## This PR contains the following types of changes

- [x] New feature (change which adds functionality)
- [x] Enhancement (improvements to existing functionality)

## Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### General

- [x] I have performed a self-review of my own code

### Code

- [x] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [x] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [x] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [x] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [x] My code changes account for null/undefined values and handle errors appropriately
- [x] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [x] My code follows the established style conventions
- [x] My code uses native material-ui components/icons/conventions when possible

### Tests

- [x] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [x] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [x] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [x] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Screenshots

Please add any relevant UI screenshots, if applicable.
Screen Shot 2021-04-23 at 12 30 39 AM
Screen Shot 2021-04-23 at 12 31 03 AM
Screen Shot 2021-04-23 at 12 31 41 AM

### CODE REVIEW

**Comment by sdevalapurkar** [comment]
can we please add a test to:
1. `AccessDenied.tsx` for line 20?
2. `RequestSubmitted.tsx` for line 18?

File: api/src/queries/administrative-activity-queries.test.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { expect } from 'chai';
+import { describe } from 'mocha';
+import { postAdministrativeActivitySQL } from './administrative-activity-queries';
+
+describe('postAdministrativeActivitySQL', () => {
+  it('Null systemUserId', () => {
```
Comment (sdevalapurkar) [inline api/src/queries/administrative-activity-queries.test.ts]: we need to add the test case for when data is null

File: app/src/layouts/AuthLayout.test.tsx [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { render } from '@testing-library/react';
+import React from 'react';
+import AuthLayout from './AuthLayout';
+import { Router } from 'react-router-dom';
+import { createMemoryHistory } from 'history';
+
+const history = createMemoryHistory();
+
+describe('AuthLayout', () => {
+  it('renders correctly', () => {
```
Comment (sdevalapurkar) [inline app/src/layouts/AuthLayout.test.tsx]: I don't think this test is doing what you expect because `context.ready` is falsey so it is not actually rendering the `props.children`, instead it is rendering a circular spinner. 

I would say, rename this test to say `renders spinner when context is not ready`
and then we can work on adding the other test case later

File: app/src/layouts/AuthLayout.test.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    const { asFragment } = render(
+      <Router history={history}>
```
Comment (sdevalapurkar) [inline app/src/layouts/AuthLayout.test.tsx]: question:
also, is the router required here?

File: app/src/pages/200/RequestSubmitted.test.tsx [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import React from 'react';
+import { Router } from 'react-router-dom';
+import RequestSubmitted from './RequestSubmitted';
+
+const history = createMemoryHistory();
+
+describe('NotFoundPage', () => {
+  it('renders correctly', () => {
+    const { asFragment } = render(
+      <Router history={history}>
```
Comment (sdevalapurkar) [inline app/src/pages/200/RequestSubmitted.test.tsx]: is the router required here?

File: app/src/pages/200/RequestSubmitted.tsx [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  return (
+    <Container>
+      <Box pt={6} textAlign="center">
+        <Icon path={mdiCheck} size={2} color="#4caf50" />
+        <h1>Access Request Submitted</h1>
+        <Typography>Your access request has been submitted for review.</Typography>
+        <Box pt={4}>
+          <Button
+            onClick={async () => {
+              window.location.href =
```
Comment (sdevalapurkar) [inline app/src/pages/200/RequestSubmitted.tsx:19]: is this the way we wanna do logouts? by manually changing the url?
It seems like there should be a function we can leverage to do this instead of changing href manually

File: app/src/pages/403/AccessDenied.tsx [hunk 6]
```diff
@@ -5,16 +5,23 @@ import { mdiAlertCircleOutline } from '@mdi/js';
...
   return (
     <Container>
       <Box pt={6} textAlign="center">
         <Icon path={mdiAlertCircleOutline} size={2} color="#ff5252" />
         <h1>Access Denied</h1>
         <Typography>You do not have permissions to access this application.</Typography>
         <Box pt={4}>
-          <Button type="submit" size="large" variant="contained" color="primary">
+          <Button
+            onClick={() => history.push('/access-request')}
```
Comment (sdevalapurkar) [inline app/src/pages/403/AccessDenied.tsx:20]: this line needs a test

File: app/src/pages/200/RequestSubmitted.tsx (hunk 5 above)
Comment (sdevalapurkar) [reply app/src/pages/200/RequestSubmitted.tsx:19]: also, this onclick needs a test case :)

File: app/src/pages/access/AccessRequestPage.test.tsx [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import { render } from '@testing-library/react';
+import { createMemoryHistory } from 'history';
+import React from 'react';
+import {
+  //MemoryRouter,
```
Comment (sdevalapurkar) [inline app/src/pages/access/AccessRequestPage.test.tsx]: remove this comment

File: app/src/pages/access/AccessRequestPage.tsx [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+      open: true
+    });
+  };
+
+  const handleSubmitAccessRequest = async (values: IAccessRequestForm) => {
+    try {
+      let response;
+
+      const accessRequestFormData = { values };
+      console.log(isSubmittingRequest);
```
Comment (sdevalapurkar) [inline app/src/pages/access/AccessRequestPage.tsx]: remove console log

File: app/src/pages/access/AccessRequestPage.tsx [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+                  below to request access.
+                </Typography>
+                <Paper elevation={2} square={true} className={classes.finishContainer}>
+                  <h2>Request Details</h2>
+                  <Box mb={3}>
+                    <form onSubmit={handleSubmit}>
+                      <Box my={3}>
+                        <Grid container spacing={3}>
+                          <Grid item xs={12}>
+                            <h3> Select which role you want to be assigned</h3>
```
Comment (sdevalapurkar) [inline app/src/pages/access/AccessRequestPage.tsx]: nit:
```suggestion
                            <h3> Select which role you want to be assigned to</h3>
```

File: app/src/pages/access/AccessRequestPage.tsx [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+              <ErrorDialog {...openErrorDialogProps} />
+            </>
+          )}
+        </Formik>
+      </Container>
+    </Box>
+  );
+};
+
+export default AccessRequestPage;
```
Comment (sdevalapurkar) [inline app/src/pages/access/AccessRequestPage.tsx:350]: this entire file is missing a lot of tests, but we can just do all of them in a separate pr 👍🏼

**Review by sdevalapurkar — CHANGES_REQUESTED** [review]

File: app/src/pages/200/RequestSubmitted.tsx (hunk 5 above)
Comment (NickPhura) [reply app/src/pages/200/RequestSubmitted.tsx:19]: The logout stuff we can refine separately, so I'd just leave this as is, or just stub it out for now.  Based on some conversations in some links Roland sent me, its possibly more involved than just calling a link (unless they've since fixed it).  But ya, eventually it would be good to add a logout function to the useKeycloakWrapper.

**Review by sdevalapurkar — APPROVED** [review]
nice work
